### PR TITLE
Support device credentials as an encryption method

### LINKED ISSUE #104: Allow PIN input next to Fingerprint to unlock DiceKey in app
I understand it is not a biometric identifier, and it implies a less secure way to access the DiceKey, but I have lots of issues with the fingerprint reader on my device (FairPhone3) where it does not recognise the fingerprint and keeps saying the "sensor is dirty" and it would help me a lot if accessing the stored DiceKey could be unlocked using a PIN.

Alternatively, could it be made possible to use a SoloKey (USB Type-C) to unlock a DiceKey in the app? Most recent phones have USB Type-C connectors.

### PR DESCRIPTION
**Pull request by angelix**
Support device credentials as a way to encrypt your DiceKey only if Biometrics is unavailable.

Fixes #104

### CODE REVIEW

File: app/src/main/java/org/dicekeys/app/encryption/AppKeystore.kt [hunk 1]
```diff
@@ -22,12 +24,15 @@ import javax.crypto.spec.IvParameterSpec
...
 
 class AppKeystore {
-    private var keyStore: KeyStore
 
-    init {
-        keyStore = KeyStore.getInstance(ANDROID_KEYSTORE).apply {
-            load(null)
-        }
+    enum class KeystoreType {
+        BIOMETRIC,
```
Comment (UppaJung) [inline app/src/main/java/org/dicekeys/app/encryption/AppKeystore.kt:29]: What do each of these mean?  Do both Face & Finger work for biometric?  How do the others work?

File: app/src/main/java/org/dicekeys/app/encryption/AppKeystore.kt [hunk 2]
```diff
@@ -79,29 +84,35 @@ class AppKeystore {
...
+                } else {
+                    builder.setUserAuthenticationValidityDurationSeconds(-1)
+                }
 
-            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
-                builder.setUserAuthenticationParameters(0, KeyProperties.AUTH_BIOMETRIC_STRONG)
-            } else {
-                builder.setUserAuthenticationValidityDurationSeconds(-1)
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
+                    builder.setInvalidatedByBiometricEnrollment(true)
```
Comment (UppaJung) [inline app/src/main/java/org/dicekeys/app/encryption/AppKeystore.kt:102]: This forces re-authentication if biometric was already used?  Please document why this is here.

File: app/src/main/java/org/dicekeys/app/encryption/AppKeystore.kt [hunk 3]
```diff
@@ -79,29 +84,35 @@ class AppKeystore {
...
-        } else {
-            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
-                builder.setUserAuthenticationParameters(0, KeyProperties.AUTH_BIOMETRIC_STRONG)
-            } else {
-                builder.setUserAuthenticationValidityDurationSeconds(-1)
+            KeystoreType.KEYSTORE -> {
+                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
+                    builder.setUserAuthenticationParameters(0, KeyProperties.AUTH_BIOMETRIC_STRONG)
+                } else {
+                    builder.setUserAuthenticationValidityDurationSeconds(-1)
```
Comment (UppaJung) [inline app/src/main/java/org/dicekeys/app/encryption/AppKeystore.kt:114]: What does this do?

File: app/src/main/java/org/dicekeys/app/encryption/BiometricsHelper.kt [hunk 4]
```diff
@@ -16,87 +21,148 @@ import org.dicekeys.dicekey.Face
...
+        return BiometricPrompt.PromptInfo.Builder().let {
+            it.setTitle(title)
+                    .setSubtitle(subtitle)
+                    .setConfirmationRequired(true)
+
+            if (keystoreType == AppKeystore.KeystoreType.BIOMETRIC) {
+                it.setAllowedAuthenticators(BiometricManager.Authenticators.BIOMETRIC_STRONG)
+                it.setNegativeButtonText(fragment.getString(android.R.string.cancel))
+            }else{
+                if(Build.VERSION.SDK_INT == Build.VERSION_CODES.R){
```
Comment (UppaJung) [inline app/src/main/java/org/dicekeys/app/encryption/BiometricsHelper.kt:156]: What special about android R?

File: app/src/main/java/org/dicekeys/app/encryption/AppKeystore.kt (hunk 2 above)
Comment (angelix) [reply app/src/main/java/org/dicekeys/app/encryption/AppKeystore.kt:102]: The documentation inside the method (cmd + click) is more that what i could have written to describe it, so i prefer to keep it clear on what each KeystoreType use.
In our case, the private key in the keystore is invalidated if a user adds a new fingerprint.

```
        /**
         * Sets whether this key should be invalidated on biometric enrollment.  This
         * applies only to keys which require user authentication (see {@link
         * #setUserAuthenticationRequired(boolean)}) and if no positive validity duration has been
         * set (see {@link #setUserAuthenticationValidityDurationSeconds(int)}, meaning the key is
         * valid for biometric authentication only.
         *
         * <p>By default, {@code invalidateKey} is {@code true}, so keys that are valid for
         * biometric authentication only are <em>irreversibly invalidated</em> when a new
         * biometric is enrolled, or when all existing biometrics are deleted.  That may be
         * changed by calling this method with {@code invalidateKey} set to {@code false}.
         *
         * <p>Invalidating keys on enrollment of a new biometric or unenrollment of all biometrics
         * improves security by ensuring that an unauthorized person who obtains the password can't
         * gain the use of biometric-authenticated keys by enrolling their own biometric.  However,
         * invalidating keys makes key-dependent operations impossible, requiring some fallback
         * procedure to authenticate the user and set up a new key.
         */
```

File: app/src/main/java/org/dicekeys/app/encryption/AppKeystore.kt (hunk 3 above)
Comment (angelix) [reply app/src/main/java/org/dicekeys/app/encryption/AppKeystore.kt:114]: It forces to re-authenticate every time, it doesn't `cache` the authorization let's say.

File: app/src/main/java/org/dicekeys/app/encryption/BiometricsHelper.kt (hunk 4 above)
Comment (angelix) [reply app/src/main/java/org/dicekeys/app/encryption/BiometricsHelper.kt:156]: From the documentation in `setAllowedAuthenticators`, some combinations are not available in older SDK.

```
             /**
             * Optional: Specifies the type(s) of authenticators that may be invoked by
             * {@link BiometricPrompt} to authenticate the user. Available authenticator types are
             * defined in {@link Authenticators} and can be combined via bitwise OR. Defaults to:
             * <ul>
             *     <li>{@link Authenticators#BIOMETRIC_WEAK} for non-crypto authentication, or</li>
             *     <li>{@link Authenticators#BIOMETRIC_STRONG} for crypto-based authentication.</li>
             * </ul>
             *
             * <p>Note that not all combinations of authenticator types are supported prior to
             * Android 11 (API 30). Specifically, {@code DEVICE_CREDENTIAL} alone is unsupported
             * prior to API 30, and {@code BIOMETRIC_STRONG | DEVICE_CREDENTIAL} is unsupported on
             * API 28-29. Setting an unsupported value on an affected Android version will result in
             * an error when calling {@link #build()}.
             *
             * <p>This method should be preferred over {@link #setDeviceCredentialAllowed(boolean)}
             * and overrides the latter if both are used. Using this method to enable device
             * credential authentication (with {@link Authenticators#DEVICE_CREDENTIAL}) will
             * replace the negative button on the prompt, making it an error to also call
             * {@link #setNegativeButtonText(CharSequence)}.
             *
             * <p>If this method is used and no authenticator of any of the specified types is
             * available at the time {@code authenticate()} is called,
             * {@link AuthenticationCallback#onAuthenticationError(int, CharSequence)} will be
             * invoked with an appropriate error code.
             *
             * @param allowedAuthenticators A bit field representing all valid authenticator types
             *                              that may be invoked by the prompt.
             * @return This builder.
             */
```

File: app/src/main/java/org/dicekeys/app/encryption/AppKeystore.kt (hunk 1 above)
Comment (angelix) [reply app/src/main/java/org/dicekeys/app/encryption/AppKeystore.kt:29]: BIOMETRIC have the most strict/secure options configured requiring Face or Fingerprint to unlock the cipher object.
AUTHENTICATION required just a device unlock, PIN, pattern etc is acceptable.

**Review by UppaJung — APPROVED** [review]

**Pull request merged**